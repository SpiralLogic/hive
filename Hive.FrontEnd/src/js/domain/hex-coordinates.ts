export type HexCoordinates = {
    readonly q: number;
    readonly r: number;
};
