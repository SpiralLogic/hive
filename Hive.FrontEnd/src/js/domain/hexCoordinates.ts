export interface HexCoordinates {
  q: number;
  r: number;
}