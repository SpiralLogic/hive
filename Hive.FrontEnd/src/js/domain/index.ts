export * from './cell';
export * from './game-state';
export * from './move';
export * from './tile';
export * from './player';
export * from './cell';
export * from './hex-coordinates';
