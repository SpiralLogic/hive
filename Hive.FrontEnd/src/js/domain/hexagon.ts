import { Tile } from './tile';
import { HexCoordinates } from './hex-coordinates';

export interface Hexagon {
    readonly coordinates: HexCoordinates;
    readonly tiles: Tile[];
}
