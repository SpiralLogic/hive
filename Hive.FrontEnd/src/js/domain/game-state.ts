import { Hexagon } from './hexagon';
import { Player } from './player';

export interface GameState {
    hexagons: Hexagon[];
    players: Player[];
}
