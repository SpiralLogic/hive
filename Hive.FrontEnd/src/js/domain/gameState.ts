import { Hexagon } from './hexagon';
import { Player, PlayerId } from './player';

export interface GameState {
  hexagons: Hexagon[];
  players: Player[];
}