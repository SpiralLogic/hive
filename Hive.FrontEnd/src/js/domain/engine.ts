import { GameState } from './game-state';
import { Move } from './move';

export interface HexEngine {
    newGame(): Promise<GameState>;
    move(move: Move): Promise<GameState>;
}
