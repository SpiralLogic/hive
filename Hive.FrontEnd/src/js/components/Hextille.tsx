import * as React from 'react';
import {Hexagon} from '../domain';
import Row from './Row';

const getWidth = (hexagons: Hexagon[]) => {
    const [min, max] = hexagons.reduce(([min, max], c) =>
        [Math.min(min, c.coordinates.q), Math.max(max, c.coordinates.q)], [0, 0]);
    return [min, max - min + 1];
};

function getHeight(sortedHexagons: Hexagon[]) {
    const firstRow = sortedHexagons[0].coordinates.r;
    const height = sortedHexagons[sortedHexagons.length - 1].coordinates.r - firstRow + 1;
    return [firstRow, height];
}


function createRows(sortedHexagons: Hexagon[]) {
    const [firstRow, height] = getHeight(sortedHexagons);
    const [firstColumn, width] = getWidth(sortedHexagons);

    const createEmptyRows = () => {
        const createEmptyRow = (i: number) => ({
            r: firstRow + i,
            row: new Array(width).fill(false),
            firstColumn
        });
        return Array.from(Array(height).keys(), createEmptyRow);
    };

    return sortedHexagons.reduce((rows, hexagon) => {
        rows[hexagon.coordinates.r - firstRow].row[hexagon.coordinates.q - firstColumn] = hexagon;
        return rows;
    }, createEmptyRows());
}

export const Hextille: React.FunctionComponent<{ hexagons: Hexagon[] }> = ({hexagons}) => {

    if (!hexagons.length) throw Error('Nothing to render !');
    const sortedHexagons = hexagons.sort((c1, c2) => c1.coordinates.r - c2.coordinates.r || c1.coordinates.q - c2.coordinates.q);
    const shiftClass = (sortedHexagons[0].coordinates.r % 2) ? 'left' : 'right';
    const rows = createRows(sortedHexagons);

    return (
        <div className={'hextille ' + shiftClass}>
            {rows.map((row) => (
                <Row key={row.r} {...row} />
            ))}
        </div>
    );
};

Hextille.displayName = 'Hextille';