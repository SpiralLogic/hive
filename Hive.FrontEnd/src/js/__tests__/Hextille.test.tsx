import { h } from 'preact';
import Hextille from '../components/Hextille';
import { HexCoordinates } from '../domain';
import { renderElement } from './helpers';

const createCell = (q: number, r: number) => ({
    coords: { q, r },
    tiles: [{ content: q + '-' + r, id: 0, playerId: 0, moves: [] as HexCoordinates[] }],
});
const createWithCells = (...coords: [number, number][]) => {
    const props: typeof Hextille.arguments.props = {
        cells: coords.map(([q, r]) => createCell(q, r)),
    };
    const hextille = renderElement(<Hextille {...props} />);
    const rows = document.body.getElementsByClassName('hex-row');
    const cells = document.body.getElementsByClassName('cell');
    const hidden = document.body.getElementsByClassName('hidden');

    return { rows, cells, hidden, hextille } as const;
};

describe('Hextille', () => {
    test('can be created with 1 cell', () => {
        const { rows, cells } = createWithCells([0, 0]);

        expect(cells).toHaveLength(1);
        expect(rows).toHaveLength(1);
    });

    test('r increases rows', () => {
        const { rows } = createWithCells([0, 0], [0, 1], [0, 2]);

        expect(rows).toHaveLength(3);
    });

    test('cells are sorted', () => {
        const { cells } = createWithCells([2, 0], [1, 0], [3, 0]);

        expect(cells[0]).toHaveTextContent('1-0');
        expect(cells[1]).toHaveTextContent('2-0');
        expect(cells[2]).toHaveTextContent('3-0');
    });

    test('rows are sorted', () => {
        const { rows } = createWithCells([0, 3], [0, 1], [0, 2]);

        expect(rows[0]).toHaveTextContent('0-1');
        expect(rows[1]).toHaveTextContent('0-2');
        expect(rows[2]).toHaveTextContent('0-3');
    });
});

describe('Row alignment', () => {
    test('shifts left when even rows have odd starting  index', () => {
        const { hextille } = createWithCells([0, 1], [0, 2], [0, 3]);

        expect(hextille).toHaveClass('left');
    });

    test('shifts right when even rows have even starting index', () => {
        const { hextille } = createWithCells([0, 2], [0, 3], [0, 4]);

        expect(hextille).toHaveClass('right');
    });
});

describe('Hextille Snapshot', () => {
    test('can move matches current snapshot', () => {
        expect(createWithCells([0, -2], [1, -2], [-1, 3], [0, 4], [5, 2], [3, 3], [1, 4])).toMatchSnapshot();
    });
});
