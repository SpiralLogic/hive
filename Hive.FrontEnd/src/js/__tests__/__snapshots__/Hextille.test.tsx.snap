// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile Snapshot can move matches current snapshot 1`] = `
Object {
  "cells": HTMLCollection [
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="0--2"
      >
        <span>
          0--2
        </span>
      </div>
    </div>,
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="1--2"
      >
        <span>
          1--2
        </span>
      </div>
    </div>,
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="5-2"
      >
        <span>
          5-2
        </span>
      </div>
    </div>,
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="-1-3"
      >
        <span>
          -1-3
        </span>
      </div>
    </div>,
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="3-3"
      >
        <span>
          3-3
        </span>
      </div>
    </div>,
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="0-4"
      >
        <span>
          0-4
        </span>
      </div>
    </div>,
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="false"
        style="--color: #85dcbc;"
        title="1-4"
      >
        <span>
          1-4
        </span>
      </div>
    </div>,
  ],
  "hextille": <div
    class="hextille right"
  >
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="0--2"
        >
          <span>
            0--2
          </span>
        </div>
      </div>
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="1--2"
        >
          <span>
            1--2
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="5-2"
        >
          <span>
            5-2
          </span>
        </div>
      </div>
    </div>
    <div
      class="hex-row"
    >
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="-1-3"
        >
          <span>
            -1-3
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="3-3"
        >
          <span>
            3-3
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="0-4"
        >
          <span>
            0-4
          </span>
        </div>
      </div>
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="1-4"
        >
          <span>
            1-4
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>
  </div>,
  "hidden": HTMLCollection [
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
    <div
      class="hidden"
    />,
  ],
  "rows": HTMLCollection [
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="0--2"
        >
          <span>
            0--2
          </span>
        </div>
      </div>
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="1--2"
        >
          <span>
            1--2
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>,
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>,
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>,
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>,
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="5-2"
        >
          <span>
            5-2
          </span>
        </div>
      </div>
    </div>,
    <div
      class="hex-row"
    >
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="-1-3"
        >
          <span>
            -1-3
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="3-3"
        >
          <span>
            3-3
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>,
    <div
      class="hex-row"
    >
      <div
        class="hidden"
      />
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="0-4"
        >
          <span>
            0-4
          </span>
        </div>
      </div>
      <div
        class="hex cell"
      >
        <div
          class="hex tile"
          draggable="false"
          style="--color: #85dcbc;"
          title="1-4"
        >
          <span>
            1-4
          </span>
        </div>
      </div>
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
      <div
        class="hidden"
      />
    </div>,
  ],
}
`;
